market_analysis/views/default.py

```diff
@@ -0,0 +1,450 @@
+    return int(shares)
+
+
+def build_stock_entry(
+        y_values, price, shares, value, max_num=None, min_num=None
+        ):
+    """Build a stock entry to be returned to build graph."""
+    return {'y_values': y_values, 'price': price, 'shares': shares,
+            'value': value, 'max': max_num, 'min': min_num}
+
+
+def build_graph(request, elements, percentage=False):
+    """Build and return the graph data from an API request for rendering."""
+    url = 'http://dev.markitondemand.com/MODApis/Api/v2/InteractiveChart/json'
+    req_obj = {     # parameters for API request
+        "parameters":
+        {
+            'Normalized': 'false',
+            'NumberOfDays': 14,
+            'DataPeriod': 'Day',
+            'Elements': elements
+        }
+    }
+    resp = requests.get(url, params=urlencode(req_obj))
+
+    if resp.status_code == 200:
+        total_shares = 0
+        total_value = 0
+        entries = {key: value for key, value in resp.json().items()}    # data from API
+        stocks = {}     # data entries for each stock
+        export = {}     # container of re-packaged data to be rendered
+
+    # build export dict for template
+        export['dates'] = format_dates(entries['Dates'])
+        export['x_values'] = entries['Positions']
+        daily_totals = [0 for j in range(len(export['x_values']))]
+
+        current_user_id = request.dbsession.query(Users).filter(
+            Users.username == request.authenticated_userid
+        ).first().id
+
+        for series in entries['Elements']:
+
+            shares = query_shares(request, current_user_id, series['Symbol'])
+            y_vals = series['DataSeries']['close']['values']
+            price = y_vals[-1]
+
+            if not shares:
+                shares = 0
+
+            total_shares += shares
+            total_value += price * shares
+
+            for i in range(len(y_vals)):
+                daily_totals[i] += (y_vals[i] * shares)
+
+            if percentage:
+                y_vals = prepare_daily_changes(y_vals)
+
+            stocks[series['Symbol']] = build_stock_entry(
+                                    y_vals, price, shares, (price * shares),
+                                    series['DataSeries']['close']['max'],
+                                    series['DataSeries']['close']['min']
+                                    )
+
+        if percentage:
+            daily_change = prepare_daily_changes(daily_totals)
+            stocks['Total'] = build_stock_entry(
+                                    daily_change, round(total_value, 2),
+                                    total_shares, round(total_value, 2)
+                                    )
+
+        export['stocks'] = stocks
+        export['total_shares'] = total_shares
+        export['total_value'] = round((total_value), 2)
+
+        return {'entry': export}
+
+    else:
+        return HTTPFound(request.route_url('api_error'))
```

Comment:
One more comment.  Everything from line 257 down to this point is functionality dedicated only to interacting wiht and manipulating the data from your stock API.  Combine that with the functionality I noted above that *should* be isolated, and you have what I would consider a nice stand-alone module for querying the api and returning the resulting data.  

Think about it like this.  If you have an app that needs to get data from an API and then manipulate that data into a certain shape, you can do as you've done here.  The effect is to staple your app to *this* api.  Or, alternatively, you can define a small API of your own for getting data shaped the right way.  Your API can be tested and proven to work.  And it can get data from the web-based API you are using now.  *Or* if that API were to go away, or you found a better one, you could alter what your API uses as a data source without needing to alter everything else about how your app works.  

This is what you have:
```
your app views ----> web api
```
But this is a superior approach:
```
your app views ----> your api facade ----> web api
```
It's superior because you can swap out web apis, or address multiple apis, or switch over to using your own database, or all of the above, and your main app views need not change at all.